### PR TITLE
Stable `all.requests`

### PR DESCRIPTION
**Pull request by max-ostapenko**
## Schema changes
- [x] `summary` column trimmed of a number of metrics (Closes https://github.com/HTTPArchive/data-pipeline/issues/149)
- [x] `payload` column trimmed of headers
- [x] `rank` data added and used for clustering (Resolves https://github.com/HTTPArchive/data-pipeline/issues/189, Closes https://github.com/HTTPArchive/data-pipeline/issues/263)
- [x] `payload` and `summary` columns are of JSON type

## Intermediate table
[`all.requests_stable`](https://console.cloud.google.com/bigquery?ws=!1m5!1m4!4m3!1shttparchive!2sall_dev!3srequests_stable)

Run reprocessing jobs with `all_requests_stable` tag.

## After reprocessing

```sql
DROP TABLE `all.requests`;

CREATE TABLE `all.requests`
COPY `all_dev.requests_stable`;
```

### CODE REVIEW

File: definitions/output/all/requests_stable.js
Comment (max-ostapenko) [inline definitions/output/all/requests_stable.js:39]:
```suggestion
```

@tunetheweb @rviscomi 
I saw couple of mentions that nobody really uses this column.
Did you discuss if we need to keep it for any future use-cases?

I would recommend to update custom_metrics and wait for a next crawl data instead.

File: definitions/output/all/requests_stable.js
Comment (max-ostapenko) [inline definitions/output/all/requests_stable.js:97]: Here is a cleaned object example:
```json
{
    "time": 0,
    "_cdn_provider": null,
    "_gzip_save": null,
    "method": "GET",
    "reqHeadersSize": 598,
    "reqBodySize": null,
    "reqCookieLen": 0,
    "status": -1,
    "respHttpVersion": "HTTP/2",
    "redirectUrl": null,
    "respHeadersSize": null,
    "respBodySize": null,
    "respSize": 0,
    "mimeType": "",
    "ext": "js",
    "type": "script",
    "format": "",
    "respCookieLen": 0,
    "expAge": 0
}
```
I think we can also remove `type`, as it's available as a separate column.

```suggestion
      ["crawlid", "firstHtml", "firstReq", "pageid", "req_accept_encoding", "req_accept_language", "req_accept", "req_if_modified_since", "req_if_none_match", "req_referer", "req_user_agent", "reqOtherHeaders", "requestid", "resp_age", "resp_cache_control", "resp_content_length", "resp_content_type", "resp_date", "resp_etag", "resp_last_modified", "resp_server", "resp_vary", "respOtherHeaders", "startedDateTime", "type", "url", "urlShort"]) as summary,
```

File: definitions/output/all/requests_stable.js
Comment (rviscomi) [reply definitions/output/all/requests_stable.js:39]: There are definitely some use cases where it's necessary to query the raw response bodies. IMO we should keep it to support those use cases, and it shouldn't affect the time/cost for any queries that don't read it directly.

Comment (tunetheweb) [reply definitions/output/all/requests_stable.js:39]: Agreed. Keep it, but massively discourage its use unless a very convincing use case is what I’ve been doing.

**Comment by max-ostapenko** [comment]
[Test run](https://console.cloud.google.com/bigquery/dataform/locations/us-central1/repositories/crawl-data-test/workspaces/dev1/workflows/1726015609-afb2868e-f285-4f9a-8f9c-fb89e09ccf68?authuser=7&project=httparchive&supportedpurview=project)



[Original](https://console.cloud.google.com/bigquery?authuser=7&project=httparchive&supportedpurview=project&ws=!1m5!1m4!4m3!1shttparchive!2sall_dev!3sdev_requests_copy) and [processed table](https://console.cloud.google.com/bigquery?authuser=7&project=httparchive&supportedpurview=project&ws=!1m5!1m4!4m3!1shttparchive!2sall_dev!3sdev_requests)

**Comment by max-ostapenko** [comment]
Would really appreciate review here, as I'm almost ready with backfill PR https://github.com/HTTPArchive/dataform/pull/10.

File: definitions/output/all/reprocess_requests.js
Comment (max-ostapenko) [inline definitions/output/all/reprocess_requests.js:44]:
```suggestion
  month >= '2022-07-01';
```
Adjust the date range before merging.

**Comment by tunetheweb** [comment]
> Would really appreciate review here, as I'm almost ready with backfill PR #10.

Hey @max-ostapenko sorry for radio silence here. I mistakenly presumed this was discussed on the last HTTP Archive maintainers call (that I missed) but just found out that @rviscomi missed this too so think we both presumed the other was dealing with this.

I think it would be good to discuss this on a call (whether the next planned one on 15th October or before). Questions/concerns I have include:
- Is this definitely all the changes we plan?
- What happens while this is running? Is the table not accessible? Or will it continue to be accessible from the run as each data is back-populated?
- How long it will take to run?
- Any impact on the next crawl (assuming it's not complete by then)?
- Any impact on Web Almanac and if we're better to wait until majority of analysis is completed there?

**Comment by max-ostapenko** [comment]
> I think it would be good to discuss this on a call (whether the next planned one on 15th October or before). Questions/concerns I have include:

I would definitely like to discuss it before the next crawl. And notify users of the upcoming changes beforehand.

> * Is this definitely all the changes we plan?

Please drop here any issues or comments that I've missed.

> * What happens while this is running? Is the table not accessible? Or will it continue to be accessible from the run as each data is back-populated?

The reprocessing will write into `all_dev.requests_stable`.

Once the reprocessing is done we can:
- run checks on the new table,
- test all the production queries changes,
- notify users with the exact date,
- update Almanac queries from this and last years,

and eventually replace the live table with `CREATE TABLE COPY` DDL. So no downtime.

> * How long it will take to run?

We can only estimate by running the first month.

In September 2024 it took 6h for requests tables without any transformations.
Let's assume it would take 2 times longer for this one - 12h.
So ~14 days.

> * Any impact on the next crawl (assuming it's not complete by then)?

Ideally we start this before the new crawl starts (so that we can already have it synced with new schema and cleaned objects to just run `INSERT INTO SELECT *`).

If anything not yet synced in WPT agent before the crawl, we can always cleanup the last crawl data within the pipeline.

> * Any impact on Web Almanac and if we're better to wait until majority of analysis is completed there?

Queries reading from struct columns need to be adjusted to a new schema to be valid.
All the analysis PRs are still pending, I'll need to go through them, and push the review where the articles were written.
Then we merge adjustments to the published queries at the moment we replace the live table. It will give a smooth experience for readers.

**Comment by tunetheweb** [comment]
We could pull `_font_details` stuff into it's own column? However I think it's niche enough (and the `payload` will be small enough with above changes) that it's OK to leave in the `payload` column. WDYT @rviscomi ?

```
    "_font_details": {
        "table_sizes": {
            "GDEF": 100,
            "GPOS": 3264,
            "GSUB": 458,
            "OS/2": 96,
...
        "counts": {
            "num_cmap_codepoints": 215,
            "num_glyphs": 238
        }
    },
```

**Comment by tunetheweb** [comment]
We need to look at native JSON columns as there were some that couldn't be processed in that and so had to use JavaScript JSON columns. See: https://github.com/HTTPArchive/httparchive.org/issues/923#issuecomment-2346828888. Is this going to be a problem if we move to native JSON columns?

**Comment by tunetheweb** [comment]
Ignore. `SAFE.PARSE_JSON(payload, wide_number_mode => 'round')` is the answer as you already implemented.

File: definitions/output/all/reprocess_requests.js
Comment (tunetheweb) [inline definitions/output/all/reprocess_requests.js:104]: Some of these are still used:

```suggestion
  JSON_REMOVE(
    SAFE.PARSE_JSON(requests.summary, wide_number_mode => 'round'),  
    '$.firstHtml',
    '$.firstReq',
    '$.req_accept_encoding',
    '$.req_accept_language',
    '$.req_accept',
    '$.req_if_modified_since',
    '$.req_if_none_match',
    '$.req_referer',
    '$.req_user_agent',
    '$.reqOtherHeaders',
    '$.requestid',
    '$.resp_age',
    '$.resp_cache_control',
    '$.resp_content_length',
    '$.resp_content_type',
    '$.resp_date',
    '$.resp_etag',
    '$.resp_last_modified',
    '$.resp_server',
    '$.resp_vary',
    '$.respOtherHeaders',
    '$.startedDateTime',
    '$.url',
    '$.urlShort'
  ) as summary,
```

**Review by tunetheweb — APPROVED** [review]
LGTM with one small request to leave the crawlids and pagesid in summary.

**Comment by rviscomi** [comment]
> We could pull `_font_details` stuff into it's own column? However I think it's niche enough (and the `payload` will be small enough with above changes) that it's OK to leave in the `payload` column. WDYT @rviscomi ?

Agreed. I could work, but I'm happy to leave it as is.

**Pull request merged**
